src/java/org/apache/cassandra/db/compaction/CompactionManager.java

```diff
@@ -135,6 +135,7 @@ protected Boolean initialValue()
     }
 
     private final CompactionExecutor executor = new CompactionExecutor();
+    private final CompactionExecutor backgroundTaskExecutor = new CompactionExecutor(1, "BackgroundTaskExecutor");
```

Comment:
With the blocking wait I suspect this will need the same number of threads as the compaction threads here -- the problem is that we now can't select new candidates for compaction until the previous tasks have all completed.

How about making this asynchronous by attaching the `BackgroundCompactionCandidate` cleanup and reschedule to the completion of the future instead of waiting?

Reply:
> With the blocking wait I suspect this will need the same number of threads as the compaction threads here
Not necessarily, a single background check may schedule many compaction tasks at once (https://github.com/datastax/cassandra/blob/ds-trunk/src/java/org/apache/cassandra/db/compaction/CompactionManager.java#L354), which actually makes the precondition for active tasks count < pool size a bit off, because after passing that check we may schedule possibly many tasks

> the problem is that we now can't select new candidates for compaction until the previous tasks have all completed
indeed, but I though that was intended logic there before as well - with the current logic, even if we allow to run `BackgroundCompactionCandidate` while there are waiting for some compactions to finish, we will immediately return because of the `hasEnoughCompactionsRunning` check

> How about making this asynchronous by attaching the BackgroundCompactionCandidate cleanup and reschedule to the completion of the future instead of waiting?
The code would definitely look better but we would still have the same problem - we would make another `BackgroundCompactionCandidate` able to start, but it would immediately return unless we changed something in `hasEnoughCompactionsRunning`

So, is your another comment the suggestion to change that logic? I'm not familiar with UCS at all, but it looks like the idea was that background compaction should not schedule tasks unless there are free slots on the executor so that when we want to enforce the compaction somehow we wouldn't have to wait possibly for very long time until all enqueued background compactions are done, am I right?

Reply:
Suppose we have 8 compaction threads and we start getting writes. After each flush, a `submitBackground` call is made. Initially they do nothing, at the fourth flush (assuming standard STCS settings) they start one task and:
- with the previous code, have 2 compaction threads busy;
- with the new code, have 1 compaction and the background thread busy.

After the next 4 flushes, in the first case the `submitBackground` call can start and schedule a concurrent compaction, but with the new code it has to wait for the first compaction to complete. The former is effectively halving the number of concurrent threads, which is bad (and something I was actually wondering about in some fallout tests). The latter is something we definitely don't want -- I don't think STCS/LCS/TWCS ever schedule more than one task (per cfs/disk/repair status) per `submitBackground` call. (UCS does this differently.)

Reply:
That's fair point, I get what you mean. Let me update the code...

Reply:
The `hasEnoughCompactionsRunning` check is a bit of a soft switch or heuristic. It's purpose is not to stop processing, but to leave it for later (where we are expected to have better information). To do this we do abort the `submitBackground` call (note that this is called quite often, on flush, on compaction completed, pediodically as a catch-all), but only if we are certain that there is some task already scheduled for this CFS.

The idea is that if we run it while we are fully busy, it might e.g. select and reserve 4 level-0 sstables to compact. Suppose by the end of a compaction, when a thread frees up, we have accumulated another 2 level-0 sstables. `submitBackground` would be called again but change nothing as the compaction is already scheduled and the 2 new sstables are not enough for a new one. If we instead cancel the first call, the second will find 6 level-0 sstables that it can compact together, which is preferable for STCS.

Reply:
This conversation makes me think that the logic here is indeed too convoluted and we need to simplify it -- probably to something like a set of CFSs that need to have their `submitBackground` called, flagged every time we get a flush or a completed compaction, and common logic that walks that set in some order every time a compaction thread frees up.

If that's something you want to do, this ticket is as good an excuse as any. If not, I think we need to switch to asynchronous cleanup and reschedule.